File: script/find_bad_positions.rb
# Usage:
#   bin/rails runner script/find_bad_positions.rb MATCH_ID BOT_NAME [--limit=8] [--tie-threshold=6] [--all]

require 'json'
require 'open3'
require 'tempfile'

match_id = ARGV.shift
bot_name = ARGV.shift

abort 'Usage: bin/rails runner script/find_bad_positions.rb MATCH_ID BOT_NAME [--limit=8] [--tie-threshold=6] [--all]' if match_id.blank? || bot_name.blank?

limit = 8
tie_threshold = 6
show_all = false

ARGV.each do |arg|
  case arg
  when /\A--limit=(\d+)\z/
    limit = Regexp.last_match(1).to_i
  when /\A--tie-threshold=(\d+)\z/
    tie_threshold = Regexp.last_match(1).to_i
  when '--all'
    show_all = true
  else
    abort "Unknown option: #{arg}"
  end
end

match = Match.find(match_id)
inspected_bot = Bot.find_by!(name: bot_name)

compiled_program =
  if match.white_player == inspected_bot
    match.white_compiled_program_snapshot
  elsif match.black_player == inspected_bot
    match.black_compiled_program_snapshot
  else
    abort "#{bot_name} is not a player in match #{match_id}"
  end

team = match.white_player == inspected_bot ? 'W' : 'B'

payload = {
  id: match.id,
  white: match.white_player.name,
  black: match.black_player.name,
  result: match.result,
  inspected_bot: inspected_bot.name,
  inspected_team: team,
  movement_notation: match.movement_notation,
  compiled_program: compiled_program,
  limit: limit,
  tie_threshold: tie_threshold,
  show_all: show_all
}

Tempfile.create(['find-bad-positions', '.json']) do |tempfile|
  tempfile.write(JSON.pretty_generate(payload))
  tempfile.flush

  command = [
    Rails.root.join('node_modules/.bin/vite-node').to_s,
    '--config',
    Rails.root.join('vitest.config.js').to_s,
    Rails.root.join('script/find_bad_positions.mjs').to_s,
    tempfile.path
  ]

  stdout, stderr, status = Open3.capture3(*command)
  print stdout
  warn stderr if stderr.present?
  exit(status.exitstatus || 1) unless status.success?
end

File: script/find_bad_positions.mjs
import fs from 'fs'

import ReplayMoveInspector from 'gameplay/replay_move_inspector'
import BotRunner from 'gameplay/bot_runner'

import { buildBoardFromNotationPrefix } from './match_analysis_common.mjs'

const payloadPath = process.argv[2]

if (!payloadPath) {
  console.error('Usage: node find_bad_positions.mjs PAYLOAD_PATH')
  process.exit(1)
}

const payload = JSON.parse(fs.readFileSync(payloadPath, 'utf8'))
const inspector = new ReplayMoveInspector({ compiledProgram: payload.compiled_program })
const runner = new BotRunner(payload.compiled_program)
const isWhite = payload.inspected_team === 'W'

function suspiciousness({ actual, tiedTopCount, trace }) {
  let score = 0
  if (actual?.score === 0) score += 3
  if (tiedTopCount >= payload.tie_threshold) score += 3
  if ((trace?.trace || []).filter(entry => entry.nodeType === 'action').length === 0) score += 2
  if (tiedTopCount > 1) score += 1
  return score
}

function reasonList({ actual, tiedTopCount, trace }) {
  const reasons = []
  if (actual?.score === 0) reasons.push('actual score 0')
  if (tiedTopCount >= payload.tie_threshold) reasons.push(`large top tie (${tiedTopCount})`)
  if ((trace?.trace || []).filter(entry => entry.nodeType === 'action').length === 0) reasons.push('no actions fired')
  if (tiedTopCount > 1 && tiedTopCount < payload.tie_threshold) reasons.push(`top tie (${tiedTopCount})`)
  return reasons
}

const flagged = []

for (let ply = 0; ply < payload.movement_notation.length; ply += 1) {
  const inspectedToMove = isWhite ? ply % 2 === 0 : ply % 2 === 1
  if (!inspectedToMove) continue

  const plyNumber = ply + 1
  const actualMoveNotation = payload.movement_notation[ply]
  const board = buildBoardFromNotationPrefix(payload.movement_notation.slice(0, ply))
  const result = inspector.inspectPosition({ board, actualMoveNotation })
  const actual = result.scoredMoves.find(r => r.key === result.actualMoveKey) || null
  const trace = actual ? runner.scoreMove({ board, moveObject: actual.moveObject, withTrace: true }) : null

  const reasons = reasonList({
    actual,
    tiedTopCount: result.tiedTopMoveKeys.length,
    trace
  })

  if (!payload.show_all && reasons.length === 0) continue

  flagged.push({
    ply: plyNumber,
    notation: actualMoveNotation,
    actualScore: actual?.score ?? null,
    topScore: result.topScore,
    tiedTopCount: result.tiedTopMoveKeys.length,
    reasons,
    suspiciousness: suspiciousness({
      actual,
      tiedTopCount: result.tiedTopMoveKeys.length,
      trace
    })
  })
}

flagged.sort((a, b) => {
  if (b.suspiciousness !== a.suspiciousness) return b.suspiciousness - a.suspiciousness
  if (b.tiedTopCount !== a.tiedTopCount) return b.tiedTopCount - a.tiedTopCount
  return a.ply - b.ply
})

console.log(`MATCH ${payload.id} ${payload.white} vs ${payload.black} ${payload.result}`)
console.log(`Flagging positions for ${payload.inspected_bot} as ${payload.inspected_team}`)
console.log('')

const rows = payload.show_all ? flagged : flagged.slice(0, payload.limit)

if (rows.length === 0) {
  console.log('No suspicious positions found for the current thresholds.')
  process.exit(0)
}

rows.forEach(row => {
  console.log(`ply ${row.ply} ${row.notation}`)
  console.log(`  score=${row.actualScore} top=${row.topScore} tied=${row.tiedTopCount} suspiciousness=${row.suspiciousness}`)
  console.log(`  reasons ${row.reasons.length > 0 ? row.reasons.join(', ') : 'none'}`)
  console.log(`  inspect  bin/rails runner script/inspect_match_moves.rb ${payload.id} "${payload.inspected_bot}" --plies=${row.ply} --all`)
  console.log('')
})

